Simplified Plugin for Better Favicon Support + Caching

DamianB-BitFlipper opened the pull request:
The plugin provided a good basis, but was not working well recently.

The favicon providers shipped with this plugin either stopped working or changed their API status. There was also a lot of overcomplication in this plugin (like choosing a random color for the circles).

I simplified the plugin and changed the favicon providers to actively maintained ones.

In order, tries to resolve:

`https://twenty-icons.com/` on the full domain
`https://twenty-icons.com/` on the root domain
`https://favicon.vemetric.com/` on the full domain
`https://favicon.vemetric.com/` on the root domain
`https://cdn.favicon0.com/` on the full domain
`https://cdn.favicon0.com/` on the root domain

Also to avoid unnecessary favicon GET requests, this plugin caches favicons in an LRU cache (maxSize 4096). If a favicon is requested that is in the cache, the data is used directly rather than hitting the network.

---

I installed this plugin locally on my mailspring and it works like a charm!

DamianB-BitFlipper committed:
feat: Remove deprecated social media avatar providers

DamianB-BitFlipper committed:
**label:** refactor

**summary:** Remove unused dependencies and simplify avatar fetching

Striffly commented:
Thank you for your contribution ! I will test it in mid-January

technobulb commented:
Thanks! Just tried it and I seem to be getting all blank icons. Any ideas?

<img width="101" height="500" alt="image" src="https://github.com/user-attachments/assets/3a22f5fa-d7cc-4af9-af4a-3860d9137872" />

DamianB-BitFlipper commented:
What do the console logs say. Open up dev tools inspect where you can see them.

technobulb commented:
```
C:\Users\user\AppDa…\error-logger.js:99 Error: Cannot find module 'C:\Users\user\AppData\Roaming\Mailspring\packages\mailspring-avatars\lib\main'
Require stack:
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package-manager.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\app-env.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\window-bootstrap.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\static\index.html
    at Package.activate (/C:/Users/appveyor/A…rc/package.ts:46:22)
    at PackageManager.activatePackage (/C:/Users/appveyor/A…ge-manager.ts:129:9)
    at /C:/Users/appveyor/A…ge-manager.ts:92:14 
Object
node:internal/modules/cjs/loader:1411 Uncaught Error: Cannot find module 'C:\Users\user\AppData\Roaming\Mailspring\packages\mailspring-avatars\lib\main'
Require stack:
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package-manager.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\app-env.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\window-bootstrap.js
- C:\Users\user\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\static\index.html
    at Package.activate (/C:/Users/appveyor/A…rc/package.ts:46:22)
    at PackageManager.activatePackage (/C:/Users/appveyor/A…ge-manager.ts:129:9)
    at /C:/Users/appveyor/A…ge-manager.ts:92:14
```
Seems like I need to build it with NPM possibly? Can you upload the already compiled plugin?
Note: I am good with computers but dont know even coding basics, but I try...

DamianB-BitFlipper commented:
Given `Cannot find module 'C:\Users\user\AppData\Roaming\Mailspring\packages\mailspring-avatars\lib\main'` it seems like the module installation broke.

I'd recommend, deleting the directory `C:\Users\user\AppData\Roaming\Mailspring\packages\mailspring-avatars` and, restarting Mailspring, and try re-installing from source.

1. Download my branch here https://github.com/DamianB-BitFlipper/mailspring-avatars/tree/main
2. Then re-install

technobulb commented:
Tried that, no luck, same error...

DamianB-BitFlipper commented:
To be honest, I think this is an issue with mailspring possibly since the module is not even getting installed. Maybe open an issue on their repo.

technobulb commented:
But when I go back and install the [old version](url=https://github.com/Striffly/mailspring-avatars), it installs fine. Trying to figure out what changes may have caused that...
Thanks for your help!

DamianB-BitFlipper commented:
Oh that’s odd. I’m not on Windows and have no way to replicate this. It installs just fine on mac.

-------- Original Message --------
On Tuesday, 12/30/25 at 10:23 technobulb ***@***.***> wrote:

> technobulb left a comment [(Striffly/mailspring-avatars#20)](https://github.com/Striffly/mailspring-avatars/pull/20#issuecomment-3698633625)
>
> But when I go back and install the old version, it installs fine. Trying to figure out what changes may have caused that...
> Thanks for your help!
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/Striffly/mailspring-avatars/pull/20#issuecomment-3698633625), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABADDDKVOFK3CAK4EMOIXG34EIZA5AVCNFSM6AAAAACP2PDYNKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMOJYGYZTGNRSGU).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

kudlav commented:
Same here after installing your branch.
Using Windows 11 25H2.
Mailspring 1.16.0

```
Error: Cannot find module 'C:\Users\Vladan\AppData\Roaming\Mailspring\packages\mailspring-avatars\lib\main'
Require stack:
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package-manager.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\app-env.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\window-bootstrap.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\static\index.html
    at Module._resolveFilename (node:internal/modules/cjs/loader:1408:15)
    at Module.a._resolveFilename (node:electron/js2c/renderer_init:2:2643)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1064:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1069:22)
    at Module._load (node:internal/modules/cjs/loader:1218:37)
    at Module.c._load (node:electron/js2c/node_init:2:18013)
    at Module.s._load (node:electron/js2c/renderer_init:2:31718)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:242:24)
    at Module.require (node:internal/modules/cjs/loader:1494:12)
    at require (node:internal/modules/helpers:135:16)
    at Package.activate (package.ts:46:22)
    at PackageManager.activatePackage (package-manager.ts:129:9)
    at PackageManager.installPackageFromPath (package-manager.ts:218:12)
    at package-manager.ts:163:14
    at AppEnvConstructor.showOpenDialog (app-env.ts:813:5) {pluginIds: Array(0)}

Uncaught (in promise) Error: Cannot find module 'C:\Users\Vladan\AppData\Roaming\Mailspring\packages\mailspring-avatars\lib\main'
Require stack:
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\package-manager.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\app-env.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\src\window-bootstrap.js
- C:\Users\Vladan\AppData\Local\Mailspring\app-1.16.0\resources\app.asar\static\index.html
    at Module._resolveFilename (node:internal/modules/cjs/loader:1408:15)
    at Module.a._resolveFilename (node:electron/js2c/renderer_init:2:2643)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1064:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1069:22)
    at Module._load (node:internal/modules/cjs/loader:1218:37)
    at Module.c._load (node:electron/js2c/node_init:2:18013)
    at Module.s._load (node:electron/js2c/renderer_init:2:31718)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:242:24)
    at Module.require (node:internal/modules/cjs/loader:1494:12)
    at require (node:internal/modules/helpers:135:16)
    at Package.activate (package.ts:46:22)
    at PackageManager.activatePackage (package-manager.ts:129:9)
    at PackageManager.installPackageFromPath (package-manager.ts:218:12)
    at package-manager.ts:163:14
    at AppEnvConstructor.showOpenDialog (app-env.ts:813:5)
```

DamianB-BitFlipper commented:
I have no way to debug this. Could one of you two please undertake investigating this issue?

My fix works on mac perfectly, so there is that to aim for ;)

technobulb commented:
Solved. Just needed to be built. 
Ideally would be set up with a Github auto-build or whatever, but for now this should work fine for you.
 Great job @DamianB-BitFlipper!
[mailspring-avatars-main BUILT.zip](https://github.com/user-attachments/files/24388668/mailspring-avatars-main.BUILT.zip)

kudlav commented:
I can confirm it works after compiling.

Ran
```
yarn
yarn build
```
Then it works. Thanks ❤️

Striffly committed:
:twisted_rightwards_arrows: Merge with main

Striffly committed:
:lock: Improve getDomain extract

Striffly committed:
:bug: Fix favicon cache race conditions

- Refactor imageToDataUrl() as a pure utility function returning base64 or URL fallback
- Fix race conditions in loadFavicon() by returning independent promises for concurrent callers
- Add proper cache cleanup on image load failures to allow retries
- Add public delete() method to LRUCache for clean API usage

Striffly committed:
:sparkles: Add TTL and memory management to LRU favicon cache

Striffly committed:
:lock: Add URL encoding for domain names in favicon URLs

Striffly committed:
:lock: Add image dimension and format validation before caching

Striffly committed:
:lock: Enforce HTTPS-only and validate data URLs before caching

Striffly committed:
:mute: Remove console logs

Striffly committed:
:lock: Add CSS injection protection for color and size props

Striffly commented:
Thank you for these changes and for this major refactoring (I must admit I never had the motivation to do this after the fork back then 😅).
I’ve made a few additional changes regarding security and cache management on top of your modifications, but as it stands, it was already an extremely high-quality piece of work. Thank you very much!
